### PR TITLE
Set current GitHub repository URL and branch into devfile projects if they're missing or didn't match

### PR DESCRIPTION
**Pull request by mshaposhnik**
### What does this PR do?
When creating a factory within an repository with devfile, the same repository must be specified in the devfile project source element. This devfile allows to omit projects part in devfile, and default
project source with given repository URL and/or branch will be applied.

### What issues does this PR fix or reference?
https://github.com/eclipse/che/issues/14547

#### Release Notes
N/A

#### Docs PR
https://github.com/eclipse/che-docs/pull/1051

### CODE REVIEW

File: wsmaster/che-core-api-factory-github/src/main/java/org/eclipse/che/api/factory/server/github/GithubFactoryParametersResolver.java [hunk 1]
```diff
@@ -115,15 +116,28 @@ public FactoryDto createFactory(@NotNull final Map<String, String> factoryParame
...
           .setProjects(
               Collections.singletonList(
                   newDto(ProjectDto.class)
                       .withSource(githubSourceStorageBuilder.buildDevfileSource(githubUrl))
                       .withName(githubUrl.getRepository())));
+    } else {
+      // update existing project with same repository, set current branch if needed
+      projects.forEach(
+          project -> {
+            if (project.getSource().getLocation().equals(githubUrl.repositoryLocation() + ".git")) {
```
Comment (sparkoo) [inline wsmaster/che-core-api-factory-github/src/main/java/org/eclipse/che/api/factory/server/github/GithubFactoryParametersResolver.java]: `.git` is not mandatory on github. will it work if I have defined project without `.git`?

Comment (sparkoo) [reply wsmaster/che-core-api-factory-github/src/main/java/org/eclipse/che/api/factory/server/github/GithubFactoryParametersResolver.java]: please add test for this usecase ^^

**Review by sparkoo — COMMENTED** [review]
does it handle forks?
what if I have more projects defined in the devfile?

File: wsmaster/che-core-api-factory-github/src/test/java/org/eclipse/che/api/factory/server/github/GithubFactoryParametersResolverTest.java [hunk 2]
```diff
@@ -156,6 +161,52 @@ public void shouldReturnFactoryFromRepositoryWithDevfile() throws Exception {
     assertEquals(factoryUrlArgumentCaptor.getValue().getDevfileFilename(), "devfile.yaml");
   }
 
+  @Test
+  public void shouldSetDefaultProjectIntoDevfileIfNotSpecified() throws Exception {
```
Comment (metlos) [inline wsmaster/che-core-api-factory-github/src/test/java/org/eclipse/che/api/factory/server/github/GithubFactoryParametersResolverTest.java:165]: I'm missing a test for setting the branch from the URL into the generated project.

**Review by metlos — COMMENTED** [review]
Took me a while to understand the reqs from the lengthy discussion at the issues but I think this works for the basic usecase of specifying a branch.

But what about the tag or commitId?

File: wsmaster/che-core-api-factory-github/src/test/java/org/eclipse/che/api/factory/server/github/GithubFactoryParametersResolverTest.java (hunk 2 above)
Comment (mshaposhnik) [reply wsmaster/che-core-api-factory-github/src/test/java/org/eclipse/che/api/factory/server/github/GithubFactoryParametersResolverTest.java:165]: i will expand this test

**Comment by mshaposhnik** [comment]
@metlos tag=branch, no differences here. About commit, it we do not support creating factories from a prticular commit ID.

File: wsmaster/che-core-api-factory-github/src/main/java/org/eclipse/che/api/factory/server/github/GithubFactoryParametersResolver.java [hunk 3]
```diff
@@ -115,15 +116,30 @@ public FactoryDto createFactory(@NotNull final Map<String, String> factoryParame
...
               Collections.singletonList(
                   newDto(ProjectDto.class)
                       .withSource(githubSourceStorageBuilder.buildDevfileSource(githubUrl))
                       .withName(githubUrl.getRepository())));
+    } else {
+      // update existing project with same repository, set current branch if needed
+      projects.forEach(
+          project -> {
+            final String location = project.getSource().getLocation();
+            if (location.equals(githubUrl.repositoryLocation())
```
Comment (sparkoo) [inline wsmaster/che-core-api-factory-github/src/main/java/org/eclipse/che/api/factory/server/github/GithubFactoryParametersResolver.java:138]: Can `source` or `location` be null? I suppose that devfile with project without location is invalid, but neither `ProjectDto` or `SourceDto` gives us guarantee that it's not null.

**Review by sparkoo — COMMENTED** [review]
Still important question for me is how it handles forks, because it's standard github workflow and we should support it. I'm not sure if it is out of the scope of this task though.

**Comment by mshaposhnik** [comment]
@sparkoo can you provide example workflow with forks? Because i don't see what do you mean.

File: wsmaster/che-core-api-factory-github/src/main/java/org/eclipse/che/api/factory/server/github/GithubFactoryParametersResolver.java (hunk 3 above)
Comment (mshaposhnik) [reply wsmaster/che-core-api-factory-github/src/main/java/org/eclipse/che/api/factory/server/github/GithubFactoryParametersResolver.java:138]: it is validated after creation so in't supposed to be invalid.

**Comment by sparkoo** [comment]
@mshaposhnik standard github flow. I have che in my space at https://github.com/sparkoo/che and I'm pushing only to branches there. Then I'm creating pull request from `sparkoo/che/branch` into `eclipse/che/master`.
With this, when we have devfile in the che repo and I start the workspace of my fork, I'd like to clone rather my fork than `eclipse/che` repo directly.

**Comment by mshaposhnik** [comment]
@sparkoo So how can we know that `sparkoo/che` is an fork of `eclipse/che` and clone it instead ?

**Comment by sparkoo** [comment]
@mshaposhnik you create workspace from `sparkoo/che` repo and there is `eclipse/che` project in the devfile so we can guess by name that it is a fork. It's still a guess and I'm not sure we should do cleverness like that.
Ideally, I would clone the fork (have it as `origin`) and add project's location as second remote.
Or it could be handled by some dialogue at workspace startup - if repo names match, ask user if it is a fork and what to clone.

**Comment by metlos** [comment]
I think there is no issue with forks actually.

The thing is that if you have devfile *without** a project, this PR will add the project from the URL you're pointing the factory on.

So if you point the factory to `eclipse/che` -  that is what's gonna be checked out. If you point it to `sparkoo/che` - the workspace will clone that...

**Comment by mshaposhnik** [comment]
crw-ci-test

**Comment by sparkoo** [comment]
@metlos yes, but what if I want to have it in the devfile? What if I want to have che devfile with `che` and `plugin-registry` sources and I want to fork that? Also this will only work with factory, so when I want to create a workspace in different way, the devfile with no project is useless.

**Comment by metlos** [comment]
So in another words, we can cover many usecases with devfile and IMHO it is an issue of educating the devfile authors:

* If you want to put the devfile in the repo of your project, so that users can use that to develop on it, don't put any projects into the devfile.
* If you want the devfile to be external to the code, you need to put the project location there, but users will have to explicitly override the project location if they want to use that devfile to work on a fork.
* If you want the devfile to aggregate multiple projects, the users will also need to provide overrides if they want to clone the projects from different locations (e.g. forks).

**Review by sparkoo — APPROVED** [review]
let's discuss forks in new issue. I'm ok with current state.

**Comment by skabashnyuk** [comment]
@mshaposhnik could you please create an improvement issue for the devfile-fork usecase in the scenario when the projects list is not empty.

**Review by skabashnyuk — APPROVED** [review]
Can you please provide an update in docs as well.

**Comment by mshaposhnik** [comment]
ci-test

**Comment by mshaposhnik** [comment]
ci-test

**Comment by artaleks9** [comment]
ci-test

**Review by metlos — APPROVED** [review]

**Pull request merged**